File: 20. Valid Parentheses/main.cpp
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>


class Solution {
 public:
    bool isValid(std::string s) {
        std::stack<char> stack;
        for (const char bracket : s) {
            if (bracket == '(' || bracket == '[' || bracket == '{') {
                stack.push(bracket);
            } else if (!stack.empty() && _brackets.at(stack.top()) == bracket) {
                stack.pop();
            } else {
                return false;
            }
        }

        return stack.empty();
    }

 private:
    const std::unordered_map<char, char> _brackets = {
        {'(', ')'},
        {'[', ']'},
        {'{', '}'},
        {')', '('},
        {']', '['},
        {'}', '{'},
    };
};


int main() {
    Solution sol;
    {
        std::string s = "()";
        std::cout << sol.isValid(s) << std::endl;
    }
    {
        std::string s = "()[]{}";
        std::cout << sol.isValid(s) << std::endl;
    }
    {
        std::string s = "(]";
        std::cout << sol.isValid(s) << std::endl;
    }
    {
        std::string s = "(";
        std::cout << sol.isValid(s) << std::endl;
    }
    return 0;
}

File: 20. Valid Parentheses/main.go
package validParentheses

var parentheses = map[rune]rune{
	'(': ')',
	'{': '}',
	'[': ']',
}

func isValid(s string) bool {
	if len(s) == 0 || len(s)%2 == 1 {
		return false
	}

	stack := make([]rune, 0, len(s)/2)
	for _, bracket := range s {
		if _, isOpening := parentheses[bracket]; isOpening {
			stack = append(stack, bracket)
		} else if len(stack) == 0 || parentheses[stack[len(stack)-1]] != bracket {
			return false
		} else {
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}
